### PR TITLE
Feature/adapter logs

### PR DESCRIPTION
**Pull request by bdmogal**
Not ready for merging yet, this is just an approach of how logging for adapters could potentially work. 

- [x] Creates a ``BatchAdapterLoggingContext`` that is used if a mapreduce program was run via a schedule that started through an adapter.
- [x] Adds the ``/namespaces/[namespace-id]/adapters/[adapter-id]/logs`` API
- [x] End-to-end testing - Deployed an adapter, used the above API to get logs for the adapter.

Some stuff that is not nice in this PR, still exploring ways to fix, but some of this may require larger refactoring
- [ ] Uses ``adapterName`` encoded in ``scheduleName``
- [ ] Separation of concerns - ``ApplicationLoggingContext`` accepts an ``adapterId``, ``BasicMapreduceContext`` knows about ``adapterId`` as well.

Things that we need in future, but are not part of this PR:
* Currently gets ``templateId``, ``programType`` and ``programId`` as parameters in the ``/namespaces/[namespace-id]/adapters/[adapter-id]/logs`` API. Have to live with this until we have an ``AdapterClient`` that can be used from the log service to get this info by making a remote call?

Build: https://builds.cask.co/browse/CDAP-DUT1346-1
Issue: [CDAP-1786](https://issues.cask.co/browse/CDAP-1786)

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java [hunk 1]
```diff
@@ -48,4 +48,6 @@
   public static final String RUN_BASE_COUNT_SIZE = "runBaseCountSize";
 
   public static final String RUN_BASE_COUNT_TIME = "runBaseCountTime";
+
+  public static final String ADAPTER_SPECIFICATION = "adapterSpeciciation";
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java]: not needed. should change to ``ADAPTER_NAME``, after moving adapter name out of schedule name

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java [hunk 2]
```diff
@@ -92,7 +97,13 @@ public BasicMapReduceContext(Program program,
     } else {
       this.userMetrics = null;
     }
-    this.loggingContext = new MapReduceLoggingContext(getNamespaceId(), getApplicationId(), getProgramName());
+    String adapterName = getAdapterName(runtimeArguments);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: Is the `runtimeArguments` the user runtime arguments? I think it should come from `ProgramOptions.getArguments`, but not `ProgramOptions.getUserArguments`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java [hunk 3]
```diff
@@ -92,7 +97,13 @@ public BasicMapReduceContext(Program program,
...
+    if (adapterName == null) {
+      this.loggingContext = new MapReduceLoggingContext(getNamespaceId(), getApplicationId(), getProgramName());
+    } else {
+      this.loggingContext = new BatchAdapterLoggingContext(getNamespaceId(), getApplicationId(), getProgramName(),
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: Why need to have a separate class just for the sake of setting the system tag? Can we just have the `AbstractLoggingContext` to have a constructor that takes `Arguments` and extract adapter name from there? Then we don't need to keep modifying code when allowing more program types to be adapter program.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/context/BatchAdapterLoggingContext.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.context;
+
+/**
+ * Logging context for instances of ETL templates
```
Comment (gokulavasan) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/context/BatchAdapterLoggingContext.java]: Is this specifically for ETL Batch Template? Or for both ETL batch and ETL realtime templates?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java (hunk 2 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: You're right, I misread this. Have fixed it to follow a similar path as ``workflowBatch``, but still need to test it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java (hunk 3 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: I was initially under the impression that adapter logs needed to be in their own partition/own directory etc. However, doesn't look like that is needed. Removed ``BatchAdapterLoggingContext``. Added an optional ``.adapterId`` tag in ``ApplicationLoggingContext``, that various programs can set if required.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/context/BatchAdapterLoggingContext.java (hunk 4 above)
Comment (bdmogal) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/context/BatchAdapterLoggingContext.java]: The initial idea I had was to have two different logging contexts. However, after exploring a bit more, removed the necessity of a new logging context. We just have an extra, optional tag in ``ApplicationLoggingContext`` now.

**Comment by bdmogal** [comment]
I have made some more progress, however, this still needs a lot of testing end-to-end. Going to do that between today and tomorrow.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java (hunk 3 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: I think it shouldn't even be set. The logging context can simply determine it from the program option. By doing so, we also enforced one way to pass in the adapter is

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: Alternatively, do the extraction at the AbstractContext level so that it can set for both metrics and logging

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java [hunk 5]
```diff
@@ -124,6 +129,14 @@ public String getWorkflowBatch() {
     return workflowBatch;
   }
 
+  /**
+   * @return the adapter name if run from within adapter, otherwise null
+   */
+  @Nullable
+  public String getAdapterName() {
```
Comment (nitinmotgi) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: What would this be when it's not part of adapter ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java [hunk 6]
```diff
@@ -106,7 +107,7 @@ public BasicMapReduceContext build(MapReduceMetrics.TaskType type,
     BasicMapReduceContext context =
       new BasicMapReduceContext(program, type, RunIds.fromString(runId), taskId,
                                 runtimeArguments, datasets, spec, logicalStartTime,
-                                workflowBatch, discoveryServiceClient, metricsCollectionService,
+                                workflowBatch, adapterName, discoveryServiceClient, metricsCollectionService,
```
Comment (nitinmotgi) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java]: Adapter is not concept known to MapReduce right ? If we make it that, what happens when it's not associated with Adapter, but used as is in more general terms.

**Comment by nitinmotgi** [comment]
Adapter is now being made as a concept similar to Application. This should not be the case.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java (hunk 5 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: ``null``, in the current design. Also, won't be added to the ``LoggingContext``, and in future, to the ``MetricsContext``, either

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java (hunk 6 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java]: when its used outside of adapters, the ``adapterName`` would be null in ``BasicMapreduceContext``. Similar pattern as ``workflowBatch``, but perhaps we need some refactoring here for ``BasicMapreduceContext`` to only receive ``Arguments`` and do the extraction that it needs within.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java [hunk 7]
```diff
@@ -63,6 +63,7 @@ public BasicMapReduceContext build(MapReduceMetrics.TaskType type,
                                      String taskId,
                                      long logicalStartTime,
                                      String workflowBatch,
+                                     String adapterName,
```
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java]: It is probably better to add new parameter in the constructor at the end rather than stick it in the middle. Make it easier to avoid merge conflict and update code.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java [hunk 8]
```diff
@@ -77,6 +79,7 @@ public BasicMapReduceContext(Program program,
                                MapReduceSpecification spec,
                                long logicalStartTime,
                                String workflowBatch,
+                               @Nullable String adapterName,
```
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: Probably better to just add new parameter to the end rather than sneak it in the middle to avoid merge conflict and update to existing code.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java [hunk 9]
```diff
@@ -126,6 +130,15 @@ public String getWorkflowBatch() {
     return jobContext.getConfiguration().get(HCONF_ATTR_WORKFLOW_BATCH);
   }
 
+  private void setAdapterName(@Nullable String adapterName) {
```
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java]: probably not a good idea to let setter to pass null. Caller to make sure not to pass null when trying to set adapter name to context config.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java [hunk 10]
```diff
@@ -178,4 +181,22 @@ public ProgramController run(Program program, ProgramOptions options) {
     }
     return controller;
   }
+
+  @Nullable
```
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Not sure what use of annotation Nullable here? It is used primarily to indicate if we could pass null as value for input parameters.
Maybe using JavaDoc explaining it could return null instead.

Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Ah looks like it could be use by IntelliJ Idea to check if method could return null.

Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: Yup, we use it pretty extensively to indicate null return values.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java (hunk 9 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java]: I could go either way on this. This same class uses two different styles of setting nulls. ``setWorkflowBatch`` allows callers to pass null and handles it internally, while ``setInputSelection`` expects callers to handle nulls before calling.

Following your suggestion, but applying the same to ``setWorkflowBatch`` as well.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java (hunk 7 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java]: Fixed. Wanted to keep it next to ``workflowBatch``, but don't feel too strongly about it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java (hunk 8 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java]: Fixed. Again, Wanted to keep it next to ``workflowBatch``, but don't feel too strongly.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java [hunk 11]
```diff
@@ -228,4 +230,16 @@ public void error(Throwable cause) {
...
+  private String getAdapterName(Arguments arguments) {
+    // TODO: Currently this logic is super ugly, should try and pass adapter name as its own separate arg
+    if (arguments.hasOption(ProgramOptionConstants.SCHEDULE_NAME)) {
+      String scheduleName = arguments.getOption(ProgramOptionConstants.SCHEDULE_NAME);
+      if (scheduleName.contains(".")) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: this won't be correct, since it is valid for a schedule name to have a dot. I see there is a todo, is that going to be fixed in a separate PR?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/AbstractProgramWorkflowRunner.java [hunk 12]
```diff
@@ -84,13 +86,15 @@ public AbstractProgramWorkflowRunner(Arguments runtimeArguments, RunId runId, Pr
    * @return a {@link Callable} of {@link RuntimeContext} for this {@link Program}
    */
   protected Callable<RuntimeContext> getRuntimeContextCallable(String name, final Program program) {
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    for (Map.Entry<String, String> systemArgument : systemArguments) {
+      builder.put(systemArgument);
+    }
+    builder.put(ProgramOptionConstants.RUN_ID, runId.getId());
+    builder.put(ProgramOptionConstants.WORKFLOW_BATCH, name);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/AbstractProgramWorkflowRunner.java]: no longer adding logical start time? is that on purpose?

Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/AbstractProgramWorkflowRunner.java]: ok i see it is added earlier

Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/AbstractProgramWorkflowRunner.java]: This is the issue where we were dropping the existing ``systemArguments`` and recreating then fresh - (lines 89-93 below). Now, workflows accept ``systemArguments`` from a higher layer (example: schedule) and only append more arguments to it, as opposed to recreating the entire ``systemArguments`` afresh. ``LOGICAL_START_TIME`` is already present in the systemArguments passed to this class from a schedule.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java [hunk 13]
```diff
@@ -80,6 +87,7 @@ public void set(BasicMapReduceContext context, CConfiguration conf,
     }
   }
 
+
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java]: nit: extra newline

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 11 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: I see, so schedules don't follow the rules for other ``Id`` classes? In that case, we should probably use a different separator for the schedule name set in ``DefaultAdapterConfigurer``. Since its not easy to get ``adapterName`` separately right now. I am unsure if that TODO will get addressed anytime in the immediate future.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/AbstractProgramWorkflowRunner.java (hunk 12 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/AbstractProgramWorkflowRunner.java]: actually, i don't see it.  where is it being set?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 11 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: yeah actually, not sure why schedule doesn't follow the rest. seems like it should.... But whatever separator is used will be a valid character so it has the same problem.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/AbstractProgramWorkflowRunner.java (hunk 12 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/AbstractProgramWorkflowRunner.java]: So if you look at ``WorkflowDriver``, it received ``ProgramOptions``. From there, it extracted ``logicalStartTime`` via ``programOptions.getArguments().get(...``. Then it passed ``logicalStartTime`` to this class via ``ProgramWorkflowRunnerFactory``. Instead of extracting there, and adding it back here in a new ``BasicArguments`` object, I just pass the original ``BasicArguments`` object around now.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java (hunk 11 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: why is it hard to get adapter? doesn't the program runner get ProgramOptions?

Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java]: So this actually follows the following class hierarchy:

*Adapter* -> *CDAP Schedule* -> *Quartz Schedule*

The ``ProgramOption``s are set in a class (``DefaultSchedulerService``) that only gets a Quartz class - ``JobExecutionContext``. And we do not pass any context information from *CDAP Schedule* classes to the Quartz Scheduler classes other than the trigger key which is an encoded string of the form: ``namespaceId:appId:programType:programId:scheduleName``.

**Comment by albertshau** [comment]
ok :+1: . let's get this in and see how we can pass the adapter name through the schedule.

**Comment by bdmogal** [comment]
Build passed. Fixed nits. Merging. Will follow up with a PR with a  different approach to passing adapter name to schedule.
